File: frontend/lib/api/payments.ts
import { apiFetch } from "@/lib/api/client";

export interface PaymentOrderCreateIn {
  amount: number;
  fee_type?: string;
  semester?: string;
  academic_year?: string;
  notes?: string;
}

export interface PaymentOrderOut {
  order_id: string;
  amount: number;
  currency: string;
  fee_type: string;
  semester: string;
  academic_year: string;
  gateway_provider: string;
  key_id?: string | null;
}

export interface PaymentVerifyIn {
  order_id: string;
  payment_id: string;
  signature?: string | null;
  payment_method: string;
  payer_details?: string | null;
}

export interface PaymentTransactionOut {
  id: string;
  student_id: string | null;
  student_email: string;
  order_id: string;
  payment_id: string | null;
  amount: number;
  currency: string;
  fee_type: string;
  semester: string;
  academic_year: string;
  status: string;
  payment_method: string | null;
  gateway_provider: string;
  receipt_no: string | null;
  receipt_hash: string | null;
  notes: string | null;
  created_at: string;
  completed_at: string | null;
}

export interface PaymentGatewayConfigIn {
  provider: "sandbox" | "razorpay" | "stripe";
  razorpay_key_id?: string;
  razorpay_key_secret?: string;
  stripe_publishable_key?: string;
  stripe_secret_key?: string;
  is_test_mode?: boolean;
  currency?: string;
}

export interface PaymentGatewayConfigOut {
  id: string;
  provider: "sandbox" | "razorpay" | "stripe";
  razorpay_key_id: string | null;
  has_razorpay_secret: boolean;
  stripe_publishable_key: string | null;
  has_stripe_secret: boolean;
  is_test_mode: boolean;
  currency: string;
  updated_at: string;
}

export interface PaymentSummaryStatsOut {
  total_collected: number;
  total_transactions: number;
  successful_count: number;
  pending_count: number;
  active_provider: string;
}

export function createPaymentOrder(
  accessToken: string,
  input: PaymentOrderCreateIn
): Promise<PaymentOrderOut> {
  return apiFetch<PaymentOrderOut>("/payments/create-order", {
    method: "POST",
    body: input,
    accessToken,
  });
}

export function verifyPayment(
  accessToken: string,
  input: PaymentVerifyIn
): Promise<PaymentTransactionOut> {
  return apiFetch<PaymentTransactionOut>("/payments/verify", {
    method: "POST",
    body: input,
    accessToken,
  });
}

export function listMyPayments(accessToken: string): Promise<PaymentTransactionOut[]> {
  return apiFetch<PaymentTransactionOut[]>("/payments/history", { accessToken });
}

export function listAllPayments(accessToken: string): Promise<PaymentTransactionOut[]> {
  return apiFetch<PaymentTransactionOut[]>("/payments/all", { accessToken });
}

export function getPaymentStats(accessToken: string): Promise<PaymentSummaryStatsOut> {
  return apiFetch<PaymentSummaryStatsOut>("/payments/stats", { accessToken });
}

export function getGatewayConfig(accessToken: string): Promise<PaymentGatewayConfigOut> {
  return apiFetch<PaymentGatewayConfigOut>("/payments/config", { accessToken });
}

export function updateGatewayConfig(
  accessToken: string,
  payload: PaymentGatewayConfigIn
): Promise<PaymentGatewayConfigOut> {
  return apiFetch<PaymentGatewayConfigOut>("/payments/config", {
    method: "POST",
    body: payload,
    accessToken,
  });
}

File: backend/app/api/v1/payments.py
"""Payment Gateway API endpoints for checkout orders, verification, and ledger records."""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, require_role
from app.db.models.user import User
from app.db.session import get_db
from app.schemas.payment import (
    PaymentGatewayConfigIn,
    PaymentGatewayConfigOut,
    PaymentOrderCreateIn,
    PaymentOrderOut,
    PaymentSummaryStatsOut,
    PaymentTransactionOut,
    PaymentVerifyIn,
)
from app.services import payment_service
from app.services.payment_service import PaymentServiceError

router = APIRouter()
require_admin = require_role("admin")
require_staff_or_admin = require_role("faculty", "admin")


@router.post("/create-order", response_model=PaymentOrderOut, status_code=status.HTTP_201_CREATED)
async def create_order(
    payload: PaymentOrderCreateIn,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PaymentOrderOut:
    """Creates a new payment gateway checkout order for fee payment."""
    try:
        return await payment_service.create_payment_order(db, current_user, payload)
    except PaymentServiceError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.post("/verify", response_model=PaymentTransactionOut)
async def verify_payment(
    payload: PaymentVerifyIn,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PaymentTransactionOut:
    """Verifies payment completion, records ledger transaction, and issues official receipt."""
    try:
        txn = await payment_service.verify_and_complete_payment(db, current_user, payload)
        return PaymentTransactionOut.model_validate(txn)
    except PaymentServiceError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc


@router.get("/history", response_model=list[PaymentTransactionOut])
async def get_my_payments(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[PaymentTransactionOut]:
    """Returns the authenticated student's fee payment history and official receipts."""
    txns = await payment_service.list_student_payments(db, current_user)
    return [PaymentTransactionOut.model_validate(t) for t in txns]


@router.get("/all", response_model=list[PaymentTransactionOut])
async def get_all_payments(
    _: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> list[PaymentTransactionOut]:
    """Admin only: Lists all student payment transactions across the university."""
    txns = await payment_service.list_all_payments(db)
    return [PaymentTransactionOut.model_validate(t) for t in txns]


@router.get("/stats", response_model=PaymentSummaryStatsOut)
async def get_payment_stats(
    _: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> PaymentSummaryStatsOut:
    """Admin only: Retrieves campus fee collections overview and gateway status metrics."""
    return await payment_service.get_payment_stats(db)


@router.get("/config", response_model=PaymentGatewayConfigOut)
async def get_gateway_config(
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> PaymentGatewayConfigOut:
    """Retrieves active payment gateway provider and public client keys."""
    return await payment_service.get_gateway_config_out(db)


@router.post("/config", response_model=PaymentGatewayConfigOut)
async def update_gateway_config(
    payload: PaymentGatewayConfigIn,
    _: User = Depends(require_admin),
    db: AsyncSession = Depends(get_db),
) -> PaymentGatewayConfigOut:
    """Admin only: Updates payment gateway provider (Sandbox, Razorpay, or Stripe) and credentials."""
    return await payment_service.update_gateway_config(db, payload)
